pinot-core/src/main/java/org/apache/pinot/core/query/optimizer/statement/JsonStatementOptimizer.java

```diff
@@ -0,0 +1,561 @@
+   * @return a Function with "AS" operator that wraps another function.
+   */
+  private static Function getAliasFunction(String alias, Function function) {
+    Function aliasFunction = new Function("AS");
+
+    List<Expression> operands = new ArrayList<>();
+    Expression expression = new Expression(ExpressionType.FUNCTION);
+    expression.setFunctionCall(function);
+    operands.add(expression);
+    operands.add(RequestUtils.createIdentifierExpression(alias));
+    aliasFunction.setOperands(operands);
+
+    return aliasFunction;
+  }
+
+  /**
+   * Example:
+   * Input : ["jsoncolumn", "x","y","z[2]"]
+   * Output: JSON_EXTRACT_SCALAR('jsoncolumn','$.x.y.z[2]','STRING','null')
+   *
+   * @param parts All the subparts of a fully qualified identifier (json path expression).
+   * @param dataType Output datatype of JSON_EXTRACT_SCALAR function.
+   * @return a Function with JSON_EXTRACT_SCALAR operator created using parts of fully qualified identifier name.
+   */
+  private static Function getJsonExtractFunction(String[] parts, DataSchema.ColumnDataType dataType) {
+    Function jsonExtractScalarFunction = new Function("JSON_EXTRACT_SCALAR");
+    List<Expression> operands = new ArrayList<>();
+    operands.add(RequestUtils.createIdentifierExpression(parts[0]));
+    operands.add(RequestUtils.createLiteralExpression(new StringLiteralAstNode(getJsonPath(parts, false))));
+    operands.add(RequestUtils.createLiteralExpression(new StringLiteralAstNode(dataType.toString())));
+
+    operands.add(RequestUtils.createLiteralExpression(getDefaultNullValueForType(dataType)));
+    jsonExtractScalarFunction.setOperands(operands);
+    return jsonExtractScalarFunction;
+  }
+
+  /**
+   * Example 1:
+   * Input : "jsonColumn.name.first = 'daffy'"
+   * Output: "JSON_MATCH(jsonColumn, '\"$.name.first\" = ''daffy''').
+   *
+   * Example 2:
+   * Input : "jsonColumn.id = 4"
+   * Output: "JSON_MATCH(jsonColumn, '\"$.id\" = 4')
+   */
+  private static void optimizeJsonPredicate(Expression expression, @Nullable TableConfig config,
+      @Nullable Schema schema) {
+    if (expression.getType() == ExpressionType.FUNCTION) {
+      Function function = expression.getFunctionCall();
+      String operator = function.getOperator();
+      FilterKind kind = FilterKind.valueOf(operator);
+      List<Expression> operands = function.getOperands();
+      switch (kind) {
+        case AND:
+        case OR: {
+          operands.forEach(operand -> optimizeJsonPredicate(operand, config, schema));
+          break;
+        }
+        case EQUALS:
+        case NOT_EQUALS:
+        case GREATER_THAN:
+        case GREATER_THAN_OR_EQUAL:
+        case LESS_THAN:
+        case LESS_THAN_OR_EQUAL: {
+          Expression left = operands.get(0);
+          Expression right = operands.get(1);
+          if (left.getType() == ExpressionType.IDENTIFIER && right.getType() == ExpressionType.LITERAL) {
+            String[] parts = getIdentifierParts(left.getIdentifier());
```

Comment:
Same here, check whether the identifier is a column before extracting the parts from it

Reply:
Fixed.